### DISCUSSION TITLE
Does glibc now compile with mold?

### DISCUSSION THREAD

**Discussion by immolo** [rui314/mold#784] [Q&A]
I see glibc has been removed from [gdoc tracker](lhttps://docs.google.com/spreadsheets/u/1/d/1LnyrUw79oY5DxBNfXGmlxr2-0n4V0AA06BzbO2cdhZg/htmlview#gid=0) but I'm still unable to get this build with the following error:

`checking if x86_64-pc-linux-gnu-gcc -m32 -mstackrealign -march=native -pipe -O2 -fuse-ld=mold -Wl,-O1 -Wl,--as-needed is sufficient to build libc... yes
checking for i686-pc-linux-gnu-nm... x86_64-pc-linux-gnu-nm
checking for python3... python3
checking version of python3... 3.11.0, ok
configure: error:
*** These critical programs are missing or too old: GNU ld
*** Check the INSTALL file for required versions.
 * ERROR: sys-libs/glibc-2.36-r4::gentoo failed (configure phase):
 *   failed to configure glibc`

Am I missing something or was the line just removed in error?

**Comment by ghost** [comment]
If glibc devs does not add support for mold it would be hard to build, as it's not something @rui314 can do on his side

**Reply by ghost** [reply]
I have the same problem as well

**Comment by rui314** [comment]
I believe the glibc's `configure` script not only checks if the linker is GNU ld but also which version it is. To simulate GNU ld, we include the phrase "compatible with GNU ld" in our `--help` message (as I discussed in my essay on software compatibility and our user-agent problem, which can be found at https://www.sigbus.info/software-compatibility-and-our-own-user-agent-problem). However, it's possible that glibc may verify the version as well. You may want to ask glibc devs to remove excessive version checks.

**Comment by SigHunter** [comment]
just as an update: this is still the case for current glibc 2.38

**Comment by satmandu** [comment]
Looking at https://github.com/bminor/glibc/blob/master/configure.ac : 
```
# Accept binutils 2.25 or newer.
libc_cv_with_lld=no
case $($LD --version) in
  "GNU gold"*)
  # Accept gold 1.14 or higher
    AC_CHECK_PROG_VER(LD, $LD, --version,
		    [GNU gold.* \([0-9][0-9]*\.[0-9.]*\)],
		    [1.1[4-9]*|1.[2-9][0-9]*|1.1[0-9][0-9]*|[2-9].*|[1-9][0-9]*],
		    LD=: critic_missing="$critic_missing GNU gold")
    ;;
  "LLD"*)
  # Accept LLD 13.0.0 or higher
    AC_CHECK_PROG_VER(LD, $LD, --version,
		    [LLD.* \([0-9][0-9]*\.[0-9.]*\)],
		    [1[3-9].*|[2-9][0-9].*],
		    LD=: critic_missing="$critic_missing LLD")
    libc_cv_with_lld=yes
    ;;
  *)
    AC_CHECK_PROG_VER(LD, $LD, --version,
		    [GNU ld.* \([0-9][0-9]*\.[0-9.]*\)],
		    [2.1[0-9][0-9]*|2.2[5-9]*|2.[3-9][0-9]*|[3-9].*|[1-9][0-9]*],
		    LD=: critic_missing="$critic_missing GNU ld")
    ;;
esac
LIBC_CONFIG_VAR([with-lld], [$libc_cv_with_lld])
```
Note that all of these matches expect to match the _beginning of the line_, and this doesn't work for matching `mold` because it only searches for the version number after matching the text `GNU ld`.

The third fallthrough option works for ld as per the resultant regex in https://github.com/bminor/glibc/blob/master/configure :
```
ld --version
GNU ld (Chromebrew) 2.44
Copyright (C) 2025 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
ld --version 2>&1 | sed -n 's/^.*GNU ld.* \([0-9][0-9]*\.[0-9.]*\).*$/\1/p'
2.44
```
However, this doesn't work for mold:
```
mold --version
mold 2.37.1 (32c8f6dd61e4b6bb92436eb67a54e22b48b54f1f; compatible with GNU ld)
mold --version 2>&1 | sed -n 's/^.*GNU ld.* \([0-9][0-9]*\.[0-9.]*\).*$/\1/p'
```

Something like this might work:
```
# Accept binutils 2.25 or newer.
libc_cv_with_lld=no
case $($LD --version) in
  "GNU gold"*)
  # Accept gold 1.14 or higher
    AC_CHECK_PROG_VER(LD, $LD, --version,
		    [GNU gold.* \([0-9][0-9]*\.[0-9.]*\)],
		    [1.1[4-9]*|1.[2-9][0-9]*|1.1[0-9][0-9]*|[2-9].*|[1-9][0-9]*],
		    LD=: critic_missing="$critic_missing GNU gold")
    ;;
  "LLD"*)
  # Accept LLD 13.0.0 or higher
    AC_CHECK_PROG_VER(LD, $LD, --version,
		    [LLD.* \([0-9][0-9]*\.[0-9.]*\)],
		    [1[3-9].*|[2-9][0-9].*],
		    LD=: critic_missing="$critic_missing LLD")
    libc_cv_with_lld=yes
    ;;
  "mold"*)
  # Accept mold 2.37.1 or higher
    AC_CHECK_PROG_VER(LD, $LD, --version,
		    [mold.* \([0-9][0-9]*\.[0-9.]*\)],
		    [2.37[1-9]*|2.3[8-9][0-9]*|2.[4-9][0-9]*],
		    LD=: critic_missing="$critic_missing mold")
    ;;
  *)
    AC_CHECK_PROG_VER(LD, $LD, --version,
		    [GNU ld.* \([0-9][0-9]*\.[0-9.]*\)],
		    [2.1[0-9][0-9]*|2.2[5-9]*|2.[3-9][0-9]*|[3-9].*|[1-9][0-9]*],
		    LD=: critic_missing="$critic_missing GNU ld")
    ;;
esac
LIBC_CONFIG_VAR([with-lld], [$libc_cv_with_lld])
```

Alternately, `mold --version` could be modified to spit out version again at the end, i.e., `mold 2.37.1 (32c8f6dd61e4b6bb92436eb67a54e22b48b54f1f; compatible with GNU ld 2.37.1)`, as that does work with the default regex:
```
echo "mold 2.37.1 (32c8f6dd61e4b6bb92436eb67a54e22b48b54f1f; compatible with GNU ld 2.37.1)" | sed -n 's/^.*GNU ld.* \([0-9][0-9]*\.[0-9.]*\).*$/\1/p'
2.37.1
```
